### PR TITLE
Add guidance on writing Dockerfiles

### PR DESCRIPTION
**Pull request by rhowe-gds**
Sorry I didn't get to finishing this, but hopefully it's useful and someone can do the final polishing?

### CODE REVIEW

**Comment by rhowe-gds** [comment]
@chrisfarms anything you'd like to add?

**Comment by issyl0** [comment]
Did you tick the button that people with push rights to this repo can also push to your fork? Then we can carry on this PR if we choose to.

**Comment by rhowe-gds** [comment]
> Did you tick the button that people with push rights to this repo can also push to your fork? Then we can carry on this PR if we choose to.

I think so

**Comment by chrisfarms** [comment]
There's some weird edge cases in concourse where resource **types** sometimes struggle with digests ...and resources themselves need to "track" a tag such as `latest` ... But these are _usually_ things you have just built and I think mentioning it here might just be confusing

**Comment by lfdebrux** [comment]
Not sure it needs any polishing?

**Comment by rhowe** [comment]
> Not sure it needs any polishing?

At the very least it needs adding to `source/index.html.md.erb`

File: source/manuals/programming-languages/docker.html.md.erb [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+i.e. instead of
+
+    FROM alpine:3.9
+
+use
+
+    FROM alpine@sha256:769fddc7cc2f0a1c35abb2f91432e8beecf83916c421420e6a6da9f8975464b6
+
+The reason for this is the non-immutability of tags. The image referenced by a
```
[the remark this answers is not in the document]
Comment (chrisfarms) [reply source/manuals/programming-languages/docker.html.md.erb]: added bit more context

File: source/manuals/programming-languages/docker.html.md.erb [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+You should specify your base image using a digest rather than a tag.
+
+i.e. instead of
+
+    FROM alpine:3.9
+
+use
+
+    FROM alpine@sha256:769fddc7cc2f0a1c35abb2f91432e8beecf83916c421420e6a6da9f8975464b6
```
[the remark this answers is not in the document]
Comment (chrisfarms) [reply source/manuals/programming-languages/docker.html.md.erb]: Added that this is about writing Dockerfiles suitable for "production"... think we should trust people to know when they can stray from best practices to improve dev workflows.

[the remark this answers is not in the document]
Comment (benthorner) [reply source/manuals/programming-languages/docker.html.md.erb]: Sure. Thanks for looking at this :-).

**Comment by rhowe** [comment]
> Perhaps something for a separate PR: in using Docker on GOV.UK we learnt a bit about how to build smaller images, by ensuring each instruction executes any cleanup commands. This might be common knowledge, but I notice it's not in the Docker docs.

I'm sure GDS (I'm unfortunately merely a consumer of GDS services these days) would appreciate additional content here.

The things to raise awareness of w.r.t. image size is to be conscious of unwanted debris left behind by the tools you use to build each layer and that layers are additive so if something is added in one layer, removing it in a subsequent layer doesn't reduce the overall image size.

Docker has an experimental `--squash` option to `docker build` which flattens layers and which can reduce image sizes.

Care should be taken with the choice of base image - it is often a balancing act of size, functionality and security support. I've found alpine to be good with the first two and OK to middling with the latter. It is also worth being aware of multi-stage builds where you can create a whole new image based on files from a previous image. `FROM scratch` is the smallest base layer you can have there!

**Pull request merged**